### DISCUSSION TITLE
Why defineTiddlerModules use `tiddler.fields.text`

### DISCUSSION THREAD

**Discussion by linonetwo** [TiddlyWiki/TiddlyWiki5#7968] [Q&A] [answered]
https://github.com/Jermolene/TiddlyWiki5/blob/0cf80e824f4dd8d557a6cb586dd3dc9ddcbc46dd/boot/boot.js#L1506-L1515

Looks like thrid param should be an object (`export` object)

**Comment by saqimtiaz** [comment]
See the comment at https://github.com/Jermolene/TiddlyWiki5/blob/0cf80e824f4dd8d557a6cb586dd3dc9ddcbc46dd/boot/bootprefix.js#L45-L49

**Reply by linonetwo** [reply]
I think most of them are `A string function body`, but I didn't see the code that eval them, do you know? Thanks.

**Reply by Jermolene** [reply, marked as answer]
They ultimately get evaluated here:

https://github.com/Jermolene/TiddlyWiki5/blob/0cf80e824f4dd8d557a6cb586dd3dc9ddcbc46dd/boot/boot.js#L898-L910

**Reply by linonetwo** [reply]
Thanks, seems `$tw.utils.evalSandboxed` needs a heavy setup to use. I'm now doing this on my plugin instead (although it needs async await):

```ts
const blob = new Blob([someJSCode], { type: 'text/javascript' });
const objectURL = URL.createObjectURL(blob);
const jsModule = await import(objectURL);
URL.revokeObjectURL(objectURL);
jsModule.default(xxx)
jsModule.somefunction(xxx)
```

**Reply by Jermolene** [reply]
That example is asynchronous. For compatibility with Node.js we need `require()` to run synchronously. I am also not certain that it will return the same object if the same code is required multiple times, which again is required by the semantics of `require()`.

But if you don't need the functionality of the core module system then as you note it is straightforward to interoperate with other systems.

**Reply by linonetwo** [reply]
> I am also not certain that it will return the same object

when use `await import` it will not. This is a feature that I'm using for https://github.com/bevyengine/bevy/discussions/11619#discussioncomment-8392977 when writing wasm based gamification plugin for Tiddlywiki.
